util/src/test/java/org/datacommons/util/ComplexValueParserTest.java

```diff
@@ -0,0 +1,166 @@
+            + "dcid: \"Years10To20\"\n"
+            + "endValue: 20\n"
+            + "name: \"Years 10 To 20\"\n"
+            + "startValue: 10\n"
+            + "typeOf: dcid:QuantityRange\n"
+            + "unit: dcid:Years\n\n";
+    assertEquals(toComplexValueMcf("[10 20 Years]"), exp);
+
+    exp =
+        "Node: Years10Onwards\n"
+            + "dcid: \"Years10Onwards\"\n"
+            + "endValue: \"-\"\n"
+            + "name: \"Years 10 Onwards\"\n"
+            + "startValue: 10\n"
+            + "typeOf: dcid:QuantityRange\n"
+            + "unit: dcid:Years\n\n";
+    assertEquals(toComplexValueMcf("[10 - dcs:Years]"), exp);
+
+    exp =
+        "Node: Years10\n"
+            + "dcid: \"Years10\"\n"
+            + "name: \"Years 10\"\n"
+            + "typeOf: dcid:Quantity\n"
+            + "unit: dcid:Years\n"
+            + "value: 10\n\n";
+    assertEquals(toComplexValueMcf("[10 Years]"), exp);
+  }
+
+  @Test
+  public void testLatLongNode() {
+    String exp =
+        "Node: latLong/3738848_-12208344\n"
+            + "dcid: \"latLong/3738848_-12208344\"\n"
+            + "latitude: \"37.3884812\"\n"
+            + "longitude: \"-122.0834373\"\n"
+            + "name: \"37.38848,-122.08344\"\n"
+            + "typeOf: dcid:GeoCoordinates\n\n";
+    assertEquals(toComplexValueMcf("[LatLong 37.3884812 -122.0834373]"), exp);
+  }
+
+  @Test
+  public void testQuantityFailure() {
+    Debug.Log log = toComplexValueFailure("[]");
+    assertTrue(log.getCounterSet().containsCounters("MCF_MalformedComplexValueString"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("malformed"));
+
+    log = toComplexValueFailure("dcs:Years 10]");
+    assertTrue(log.getCounterSet().containsCounters("MCF_UnenclosedComplexValue"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("not enclosed in brackets"));
+
+    log = toComplexValueFailure("[Years 10");
+    assertTrue(log.getCounterSet().containsCounters("MCF_UnenclosedComplexValue"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("not enclosed in brackets"));
+
+    log = toComplexValueFailure("[Years 1 2 3]");
+    assertTrue(log.getCounterSet().containsCounters("MCF_MalformedComplexValueParts"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("2 or 3"));
+
+    log = toComplexValueFailure("[dcs:Years -]");
+    assertTrue(log.getCounterSet().containsCounters("MCF_QuantityMalformedValue"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("must be a number"));
+
+    log = toComplexValueFailure("[dcs:Years - -]");
+    assertTrue(log.getCounterSet().containsCounters("MCF_QuantityRangeMalformedValues"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("must be a number"));
+
+    log = toComplexValueFailure("[a b Years]");
+    assertTrue(log.getCounterSet().containsCounters("MCF_QuantityRangeMalformedValues"));
+    assertTrue(log.getEntries(0).getUserMessage().contains("must be a number"));
```

Comment:
maybe also add a test here for "[ years 11 2 ]" (are spaces before/after the square bracket allowed)?

Reply:
Good point. Yes, spaces before/after are okay. I've added this example in the success testcase above.